feat: 2.2 new srp oauth service

ieow opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**
Integration the following social login via the OAuth Service

Provider supported
Google
Apple

Proposed Authentication Flow using seedless onboarding
![Screenshot 2025-04-29 at 8 25 46 PM](https://github.com/user-attachments/assets/99f93c54-ebae-4ac9-a62c-069c4c09beb4)

This PR mainly integrate the OAuthServices integration into Metamask.
It does not include the Login UI Flow.
The Login UI Flow will be added on top of the new SRP UI Flow


This PR we are removing the code fence in order to be in parity with the extension where it uses runtime env flag instead.


<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

ieow commented:
@SocketSecurity ignore expo-auth-session@6.0.3

####  expo-auth-session@6.0.3
This package is being used to execute secure oauth2.0 login

ieow committed:
feat: clean up oauth service
remove unwanted lint changes

ieow committed:
fix: remove build flag

ieow committed:
fix: remove unused added assets

ieow committed:
fix: oauth config

ieow committed:
chore: address comment

ieow committed:
fix : check for getBundleId

ieow committed:
fix: default ios google client bundleid

ieow committed:
fix: remove oauth env from js.env.example

FrederikBolding reviewed (dismissed):
Code fence change LGTM

chaitanyapotti commented on app/core/OAuthService/OAuthLoginHandlers/baseHandler.ts:
```diff
@@ -0,0 +1,156 @@
+ * @param input - The base64 encoded string to pad
+ * @returns The padded string
+ */
+function padBase64String(input: string) {
```
use the new fn without buffer. copy from extension

ieow replied:
Extension is using buffer as well
https://github.com/MetaMask/metamask-extension/blob/9f74d4fb3ae91e35d524a572f493365f4183d0da/app/scripts/services/oauth/utils.ts#L7

ieow replied:
replace with `fromBase64UrlSafe`

ieow replied:
address in [be62797](https://github.com/MetaMask/metamask-mobile/pull/15877/commits/be627973e91962e15211163ef990418d0d577458)

chaitanyapotti commented on app/core/OAuthService/OAuthLoginHandlers/androidHandlers/apple.ts:
```diff
@@ -0,0 +1,148 @@
+      clientId: this.clientId,
+      random: this.nonce,
+    });
+    const authRequest = new AuthRequest({
```
use bff flow

chaitanyapotti commented on app/core/OAuthService/OAuthLoginHandlers/config.ts:
```diff
@@ -0,0 +1,96 @@
+
+export const OAUTH_CONFIG: Record<string, Record<string, string>> = {
+    development: {
+      ANDROID_APPLE_CLIENT_ID: 'com.web3auth.appleloginextension',
```
move client ids to env variables

ieow replied:
[5f0d3e6](https://github.com/MetaMask/metamask-mobile/pull/15877/commits/5f0d3e619851eaf0398f789bdf8d5873cd528041)

chaitanyapotti commented on app/core/OAuthService/OAuthLoginHandlers/iosHandlers/google.ts:
```diff
@@ -0,0 +1,102 @@
+  public readonly OAUTH_SERVER_URL =
+    'https://accounts.google.com/o/oauth2/v2/auth';
+
+  readonly #scope = ['email', 'profile'];
```
add openid
```suggestion
  readonly #scope = ['email', 'profile', 'openid'];
```

ieow replied:
address in [be62797](https://github.com/MetaMask/metamask-mobile/pull/15877/commits/be627973e91962e15211163ef990418d0d577458)

ieow committed:
fix: address comment

ieow committed:
fix: extension parity
update config
apple bff

ieow committed:
chore: move client id out of config

ieow committed:
fix: update acm with main commit

ieow committed:
fix: mock base handler

ieow committed:
fix: tests

ieow committed:
fix: push coverage

ieow committed:
fix: add test for getAuthToken

ieow committed:
Merge branch 'feat/new-srp-oauth-service' into feat/oauth-clientId

ieow committed:
feat: add

tommasini commented on app/core/OAuthService/OAuthLoginHandlers/androidHandlers/apple.ts:
```diff
@@ -0,0 +1,202 @@
+import { ResponseType, AuthRequest } from 'expo-auth-session';
```
Also weird the folder structure choosen, the apple file is inside the androidHandlers folder, does this make sense?

ieow replied:
We have both apple login and google login in android device
and apple login and google login in ios device

tommasini commented on app/core/OAuthService/OAuthLoginHandlers/baseHandler.ts:
```diff
@@ -0,0 +1,269 @@
+    challenge: string;
+  } {
+    const codeVerifier = randomBytes(32);
+    const challenge = sha256(codeVerifier);
```
Can we replace this hash for a native one?
https://github.com/margelo/react-native-quick-crypto?tab=readme-ov-file#usage

tommasini commented on package.json:
```diff
@@ -368,6 +370,7 @@
     "react-native-fs": "^2.20.0",
     "react-native-gesture-handler": "^2.25.0",
     "react-native-get-random-values": "^1.8.0",
+    "@metamask/react-native-acm": "git+https://github.com/MetaMask/react-native-google-acm.git#7a9c4ce54c97b786a5f2cd869b3b3c99f5079bdd",
```
Can we publish a version for this instead?

ieow replied:
@metamask/react-native-acm seems cannot be release by today
Could we go in with github link for now?
I will update with released version once it is release on 2.3

tommasini commented on app/core/OAuthService/OAuthLoginHandlers/androidHandlers/apple.ts:
```diff
@@ -0,0 +1,202 @@
+  protected appRedirectUri: string;
+
+  get authConnection() {
+    return AuthConnection.Apple;
```
Hmm I don't think that's true, with that getter, if I understand correctly we are returning "apple", which is what is defined at the OAuthInterface file, am I missing something? How will this getter return something different than "apple"?

chaitanyapotti replied:
this handler is a sub class for `BaseLoginHandler` which defines the `LoginHandler` interface
The idea is that each handler has a getter which defines the type of handler it is.!

ieow committed:
fix: address comment
update with quick crypto lib

ieow commented on app/core/OAuthService/OAuthLoginHandlers/baseHandler.ts:
```diff
@@ -0,0 +1,269 @@
+    codeVerifier: string;
+    challenge: string;
+  } {
+    const codeVerifier = randomBytes(32);
```
address in commit [752057d](https://github.com/MetaMask/metamask-mobile/pull/15877/commits/752057db9d3ede2adfb6a30eb26ec2d3727325ba)

ieow committed:
fix: test

ieow committed:
fix: mock react-native-quick-crypto

ieow committed:
fix: lint

ieow committed:
chroe: remaping buildtype env

ieow committed:
Merge branch 'feat/new-srp-oauth-service' into feat/oauth-clientId

ieow committed:
Feat/oauth-clientId (#16823)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**
As requirement to match parity with extension, the client Id need to be
move out from config to env

This pr delete the client ids in the config file and add remapping of
the env variable for the buildtype


<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

ieow commented on app/core/OAuthService/OAuthLoginHandlers/baseHandler.ts:
```diff
@@ -0,0 +1,156 @@
+ * Base class for the login handlers
+ */
+export abstract class BaseLoginHandler {
+  public nonce: string;
```
address in [be62797](https://github.com/MetaMask/metamask-mobile/pull/15877/commits/be627973e91962e15211163ef990418d0d577458)

ieow commented on app/core/OAuthService/OAuthLoginHandlers/androidHandlers/google.ts:
```diff
@@ -0,0 +1,86 @@
+ * AndroidGoogleLoginHandler is the login handler for the Google login on android.
+ */
+export class AndroidGoogleLoginHandler extends BaseLoginHandler {
+  readonly #scope = ['email', 'profile'];
```
address in [be62797](https://github.com/MetaMask/metamask-mobile/pull/15877/commits/be627973e91962e15211163ef990418d0d577458)

ieow committed:
Merge branch 'main' into feat/new-srp-oauth-service

ieow committed:
fix: bff start with authserver url
default js.env.example for seedless-onboarding set to false

ieow committed:
fix: use base64 url for pcke

ieow committed:
fix: update with mm authserver flow
fix apple bff

ieow committed:
chore: update example with mm dev client id

chaitanyapotti committed:
fix code class

ieow committed:
fix: bff and typing

ieow committed:
fix: remapping prod variable for seedless onboarding

ieow committed:
fix: tests

ieow committed:
fix: lint

ieow committed:
fix: update with qa env checking for buildtype

ieow committed:
fix: js.env.example

ieow committed:
fix: lint

chaitanyapotti committed:
Merge branch 'main' into feat/new-srp-oauth-service

hieu-w committed:
Merge branch 'main' into feat/new-srp-oauth-service

hieu-w committed:
test: fix flaky scroll to element util

hieu-w committed:
Merge branch 'main' into feat/new-srp-oauth-service

hieu-w committed:
test: revert ImportWalletView page object to latest

chaitanyapotti committed:
Merge branch 'main' into feat/new-srp-oauth-service

chaitanyapotti commented:
cursor comments reg. missing oauth remaps are intentional. we will update in subsequent pr when the prod server is up

hieu-w committed:
Revert "test: fix flaky scroll to element util"

This reverts commit 7b7db7bfe7a1f64fb510018ba5031209ef03f7e6.

tommasini commented:
Adding the label to skip e2e since we have a very specific platform wise e2e test failing across multiple branches, it seems that we are timing out the platform tests due to a test not identified yet that doesn't stop runs.
